Add daemon.json with overlay2 to fix container startup issue

polmoorx opened the pull request:
#### Merge Checklist  <!-- REQUIRED -->
**All** boxes should be checked before merging the PR
- [x] The changes in the PR have been built and tested
- [] cgmanifest file has been updated if required
- [x] Ready to merge

#### Description <!-- REQUIRED -->
<!-- Please include a summary of the changes and the related issue. List any dependencies that are required for this change. -->
On EMT OS, when we tried to start Docker, it failed because the default overlay2 storage driver requires a writable, overlay-compatible filesystem, but /var/lib/docker resides under /var, which on EMT OS can be mounted as read-only or which isn’t compatible with overlay operations. This prevented Docker from launching. To fix this, we configured Docker to store its data in /opt/docker-data, a persistent and writable location that supports overlay2, by adding a daemon.json file and updating the RPM spec to create the required directories, install the configuration, and safely migrate any existing data from /var/lib/docker using rsync. With this change, Docker now starts and runs reliably on both RT and non-RT EMT OS, using a stable and compatible storage location.
<!-- Fixes # (issue) -->

#### Any Newly Introduced Dependencies  <!-- REQUIRED -->
No
#### How Has This Been Tested? <!-- REQUIRED -->
Tested manually.

yepmunchun commented on SPECS/moby-engine/moby-engine.spec:
```diff
@@ -103,11 +104,23 @@ mkdir -p %{buildroot}%{_unitdir}
 install -p -m 644 %{SOURCE2} %{buildroot}%{_unitdir}/docker.socket
 
+mkdir -p %{buildroot}/opt/docker-data
+chmod 0700 %{buildroot}/opt/docker-data
```
Redundant to line 118-119, where similar `mkdir -p /opt/docker-data` and `chmod 0700` also added there.

polmoorx replied:
Thank you for the comment. I have removed the redundant mkdir -p /opt/docker-data and chmod 0700 /opt/docker-data from the %install section, keeping only the creation and permission commands in the %post script.

This is because:

- The %install section runs during package build time, preparing the package filesystem (buildroot).

- The %post section runs during package installation on the live system, which is the correct place to ensure /opt/docker-data exists with proper permissions on the target device.

yepmunchun commented:
Before switching `data-root` to new location, have you tried to add `/var/lib/docker` in `layout.env`'s `PERSISTENT_BIND_PATHS` ?

polmoorx commented:
> Before switching `data-root` to new location, have you tried to add `/var/lib/docker` in `layout.env`'s `PERSISTENT_BIND_PATHS` ?

We did try adding /var/lib/docker to PERSISTENT_BIND_PATHS before changing the data-root to /opt/docker-data. But because on EMT OS /var is often temporary (tmpfs) early in boot, Docker’s data under /var/lib/docker wasn’t actually persistent, so the overlayfs errors continued.

Changing Docker’s data-root to /opt/docker-data and ensuring it is persistent and created at runtime fixed the problem.

yepmunchun commented on SPECS/moby-engine/docker.service:
```diff
@@ -11,6 +11,7 @@ Type=notify
 # the default is not to use systemd for cgroups because the delegate issues still
 # exists and systemd currently does not support the cgroup feature set required
 # for containers run by docker
+ExecStartPre=/usr/bin/mkdir -p /opt/docker-data
```
There are multiple places implemented similar action of this directory creation, e.g. post package installation in %post, packages files directory in %files. Why do we still need run this every time when docker.service start?

Jkintalix replied:
Addressed

yepmunchun commented on SPECS/moby-engine/docker.service:
```diff
@@ -44,4 +45,4 @@ Delegate=yes
 [Install]
-WantedBy=multi-user.target
\ No newline at end of file
+WantedBy=multi-user.target
```
no changes, revert this line.

Jkintalix replied:
Addressed

yepmunchun commented on toolkit/imageconfigs/additionalconfigs/layout.env:
```diff
@@ -49,6 +49,10 @@ PERSISTENT_BIND_PATHS+="
+# Docker persistent data directory
+PERSISTENT_BIND_PATHS+="
+  /opt/docker-data"
+
```
Please test if data in `/opt/docker-data` able to persists over reboot without this change.

Jkintalix replied:
Addressed

yepmunchun commented on SPECS/moby-engine/moby-engine.spec:
```diff
@@ -118,10 +129,16 @@ fi
 %license LICENSE NOTICE
 %{_bindir}/dockerd
 %{_libexecdir}/docker-proxy
+%dir /opt/docker-data
```
Post package installation will handle `/opt/docker-data` creation, if this is not needed, please remove.

Jkintalix replied:
Addressed

polmoorx committed:
add daemon.json with overlay2 to fix container startup issue

Added a daemon.json file to set "overlay2" as the storage driver
and moved Docker’s data directory to /opt/docker-data. This change
makes sure Docker starts properly on EMT OS.

Also updated the spec file to install this config and copy any existing
Docker data to the new location.

Signed-off-by: Polmoorx Shiva Kumar <polmoorx.shiva.kumar@intel.com>

andy-vm commented on SPECS/moby-engine/moby-engine.spec:
```diff
@@ -104,10 +105,19 @@ mkdir -p %{buildroot}%{_unitdir}
+mkdir -p /opt/docker-data
+chmod 0700 /opt/docker-data
+
+if [ -d /var/lib/docker ] && [ ! -d /opt/docker-data/image ]; then
```
this is new installation, there is nothing to migrate before starting docker daemon, please test it out

polmoorx replied:
Hi @andy-vm,

ITEP-70133: for the overlay issue, we understand that for a fresh installation, there’s no existing /var/lib/docker data to migrate. The rsync step is included only as a safeguard, it runs harmlessly even if the source directory is empty or doesn’t exist.
We’ve verified that the fix works on new installations, the directory setup and Docker daemon startup complete successfully without errors.

andy-vm replied:
then why we need rsync if there is no existing /var/lib/docker data?

polmoorx replied:
Hi @andy-vm 
 
Thank you for the clarification, you’re absolutely right that on a fresh installation, there’s no existing /var/lib/docker data to migrate.
 
The rsync step was added mainly as a safety measure for **upgrade or reinstallation** cases where /var/lib/docker might already contain data from a previous setup or manual configuration. It runs only if that directory exists, so on new installations it effectively does nothing, but helps ensure backward compatibility and data preservation in other scenarios.
 
We’ve verified that on fresh EMT OS installs, the directory setup and Docker daemon start successfully without any migration taking place.

andy-vm replied:
during reinstallation, the whole OS is overwritten, nothing left to rsync.

Is there any use case that we need to re-install single RPM during upgrade?

raw image is immutable

polmoorx replied:
Hi @andy-vm ,

Thanks for the clarification. Since the OS image is immutable, there’s no need for rsync.
I’ve removed that part from the patch and kept only the /opt/docker-data setup.
Tested on fresh EMT OS, and the overlay issue is resolved. Docker starts and works fine now.

Please find the reference log snippets:
<img width="498" height="184" alt="image" src="https://github.com/user-attachments/assets/6ea7d1b1-29c6-42e4-ae97-2030ff1e4bb4" />
<img width="369" height="162" alt="image" src="https://github.com/user-attachments/assets/afdb1a30-0cc9-4abc-8af1-1e48e557699b" />
<img width="629" height="422" alt="image" src="https://github.com/user-attachments/assets/78117c6a-abab-435d-8930-ec091c94cee3" />

andy-vm replied:
since we customized the rpm spec, please fix the lint by updating Vendor and Distribution

<img width="1493" height="615" alt="image" src="https://github.com/user-attachments/assets/7f4f6118-7f90-4adb-8314-a3093e3c7369" />

polmoorx replied:
Hi @andy-vm, 

Updated the vendor name and distribution name..

andy-vm reviewed (approved):
LGTM

aaroncyew commented:
Automated Messages: Label 'moby-engine' has been added to this Pull Request.

aaroncyew commented:
Automated Messages: Label 'moby-engine' has been added to this Pull Request.

aaroncyew commented:
Automated Messages: Label 'moby-engine' has been added to this Pull Request.

aaroncyew commented:
Automated Messages: Label 'moby-engine' has been added to this Pull Request.

aaroncyew commented:
Automated Messages: Label 'moby-engine' has been added to this Pull Request.

aaroncyew commented:
Automated Messages: Label 'moby-engine' has been added to this Pull Request.

aaroncyew commented:
Automated Messages: Label 'moby-engine' has been added to this Pull Request.

aaroncyew commented:
Automated Messages: Label 'moby-engine' has been added to this Pull Request.

aaroncyew commented:
Automated Messages: Label 'moby-engine' has been added to this Pull Request.
